### ISSUE TITLE
Awkward dots definition syntax

### ISSUE THREAD

**Issue by aaronmars** [0xfe/vexflow#310]
As far as I can tell, when specifying dots on a note, they must be handled in 2 places.  Even then, there is an inconsistent result.

The following code produces a StaveNote:
```javascript
var n = new Vex.Flow.StaveNote({ keys: ["f/5"], duration: "16", dots: 1});
n.dots === 1; //true
n.getIntrinsicTicks(); //1536
```
...but there is no dot visible when the note is rendered.  In order to get that to happen, you have to do:
```javascript
var n = new Vex.Flow.StaveNote({ keys: ["f/5"], duration: "16" });
n.addDotToAll();
n.dots === 1; //true
n.getIntrinsicTicks(); //1024
```
This is inconsistent.  By adding dots in the note constructor, no dot is rendered, but the dot count and note ticks are correct.  By adding dots after the fact (with `addDotToAll()`), the dot is rendered, the dot count is incremented, but the note ticks do not change.

I'm wondering what the desired fix is here.  My first impulse is to have the constructor-supplied `dots` parameter behave the same as the a subsequent call to `addDotToAll()`:
* When dots are defined in the constructor, the dot count and ticks will be correct, and a dot will be rendered.
* When addDotToAll() is called, the dot count will be incremented, the tick count will get incremented, and the added dot will be rendered.

The final bit of confusion is the fact that `addDot(index)` is possible... What is the desired result of `note.dots` and `note.getIntrinsicTicks()` when a dot is added to only one note in the voice?

**Comment by 0xfe** [comment]
Yeah, this sucks. We need a cleaner way to do this (and not require more than a single call.) Usually, I try to keep the drawing calls separate from the semantic calls, mainly because a note can appear one way in isolation, but be played differently (e.g., hidden accidentals implied on a stave.)

But here I don't think there's ever a case where a note can have a dot, and not share its timing semantics.

For `addDot(index)`, I think the correct value for `getIntrinsicTicks()` should be the maximum tick value of the note.

**Comment by aaronmars** [comment]
I don't think it's so bad to have a constructor option for `Note` that adds the modifiers and adjusts the ticks.  Although, in general, I do support your efforts to keep drawing stuff away from math stuff.

I'm wondering if `addDot()` and `addDotToAll()` should be moved to `Note` (rather than `StaveNote`)?  The dot count (`this.dots`) and the ticks (through `Vex.Flow.parseNoteData()`) are managed there already.  In addition to that change, I'd do the following:

In `Note.init()`, add dot modifiers based on the passed-in dots count.  This allows for dots to be drawn through just the constructor parameter.  No `addDot[ToAll]()` required.

`Note.addDot(index)` would adjust the ticks, but only if the dot being added would create a new maximum.

**Comment by aaronmars** [comment]
OK.  Nevermind that first bit.  The functions should stay in `StaveNote`.  `Note` has no concept of keys, and that's sort of the whole point of `addDotToAll()`.

**Comment by 0xfe** [comment]
Thanks Aaron. So, after going through the code, here are my thoughts:

* The `dots` parameter is unfortunately named. It exists strictly for augmenting the duration of the note. It should probably have been called something else (or be lumped into the `duration` parameter.)

* I'm not a big fan of any of the `add*()` methods, except `addModifier()` -- they way too tightly couple specific modifiers with the note. It's yucky to have to update `StaveNote` everytime you add a new modifier. I'd much rather have static helper functions for stuff like that. (A separate builder API as mentioned in #195 would be ideal.)

* I'm not a fan of adding modifiers in the constructor for the same reason.

* Ticks must either be initialized during construction, or explicitly with `setDuration()`. 

* This is unfortunately the state of the world today, and I'd prefer to remain backwards compatible until we have a clear plan for a new API. Meanwhile we can deprecate some of these ugly bigs in the comments (and probably add console warnings based on a flag.)

So , moving forward, I propose the following:

1) Deprecate `dots` in the constructor and alias it to some new clearer parameter name (`dottedDuration` maybe.)
2) No tick adjustment on adding any modifiers.
3) Move all duration setting and duration helper methods that are currently in `tables.js` up to `Tickable` (which is `Note`'s base class), so all the tick calculations/parsing is done in one place.
4) Deprecate all the `add*()` methods (except `addModifier*()`).

I did some quick grepping and it seems like this shouldn't be too hard -- there are only about 6 files which do duration specific stuff (each in a small number of places.) So it shouldn't be too hard to rationalize and clean up this stuff.

**Comment by aaronmars** [comment]
OK.  So, after these items are addressed, the goal is to have the One True Way<sup>TM</sup> of adding dots to be in the duration string, right?  Starting with the code in the original issue, we want the following to work:

```javascript
var n = new Vex.Flow.StaveNote({ keys: ["f/5"], duration: "16d" });
n.addModifier(0, new Vex.Flow.Dot());
```
...where the 'd' in the constructor's duration param handles the ticks (math stuff), and the `addModifier()` call handles the rendering.

It does get verbose when multiple dots are desired, or when rendering a chord:
```javascript
var n = new Vex.Flow.StaveNote({ keys: ["f/5"], duration: "16dd" });
n.addModifier(0, new Vex.Flow.Dot());
n.addModifier(0, new Vex.Flow.Dot());

var n = new Vex.Flow.StaveNote({ keys: ["c/5", "f/5"], duration: "16d" });
n.addModifier(0, new Vex.Flow.Dot());
n.addModifier(1, new Vex.Flow.Dot());
```
Not that I'm complaining... this is just full disclosure.  The TBD builder API would presumably provide shortcuts, if deemed necessary.  I do prefer the verbosity at this low level of the lib.

If this seems good, I'll also sanitize the tests to all do it this way.  They're often cited as reference implementations for people using the library, so it would be good to only have the non-deprecated path represented there.

Finally, I think it might be good to leave everything pretty much as-is.  This, along with the deprecation warnings, would be the best way to protect backward compatibility.  As a first, more incremental step, I'd propose the following changes:

1. Deprecation warnings for all `add*()` functions (except `addModifier()`).
2. Deprecation warning if `dots` is supplied in the `Note` (and, by extension, `StaveNote`) constructor .
3. Refactor of tests to conform to the desired pattern of handling dots.

**Comment by 0xfe** [comment]
Yes, sounds great. My only nitpick would be that replacing `dot` with `durationDots` might be easier to maintain than overloading the duration string -- but yes, everything else sounds good.

Also, I agree with you about verbosity: it's fine. Higher level builder APIs or simple helper functions can be used to get around it.

Thanks for opening this issue, and offering to implement the next steps. Your proposal sounds great -- please remember to also update `docs/tutorial.html` with the new calls.

**Comment by aaronmars** [comment]
Not a problem at all!

I'm totally cool with aliasing `dots` to `durationDots` (probably `duration_dots`, to stick with the code style).  It's more programmatic than overloading the duration string, for sure.  In that case, should the 'd' part of the duration string trigger the deprecation warning?  The other alternative would be to have them be additive, but that goes against having only one (non-deprecated) way of initializing a note's dots.

In other words, is this desirable usage?
```javascript
var n = new Vex.Flow.StaveNote({ keys: ["f/5"], duration: "16d", durationDots: 1 });
n.addModifier(0, new Vex.Flow.Dot());
n.addModifier(0, new Vex.Flow.Dot());
```

**Comment by mscuthbert** [comment]
My experience in playing with representation and "ticks" in `music21` was that it was most helpful for users to have two genera of durations, linked and unlinked, with linked being the default.  A linked duration got all its information about ticks (quarterLength in m21) from its duration type (16th, eighth, etc.), number of dots, and tuplets. Changing the ticks changed the type, dots, tuplets, while changing type, dots, tuplets changed the ticks, UNLESS the two were set as unlinked (which can be helpful for some engraving situations).  All note keys (pitches in m21) shared the same duration object by default (though refs and weak refs) but they could have their own duration references set manually for cases like a chord where the top note was represented as a half note and the rest as quarters. 

I don't think that this is necessarily THE WAY to do things for all software, or for Vexflow.  But some things learned from that may be: change the duration key/value to accept an object representing everything about the StaveNote's duration:

```javascript
    var n = new Vex.Flow.StaveNote({ keys: ["f/5", "g/5"], duration: { type: "16", dots: 1 }  })
```

This would seem to allow a more backward compatible mechanism simply by checking if duration's value is a string or non-null object.

**Comment by 0xfe** [comment]
Thanks Michael. I think here changing duration to an object seems like a great idea, and would make it slightly more extensible.

Aaron, yes, I think "d" should trigger a deprecation warning -- it's a bit too hacky. Meanwhile, definitely not additive; that would probably be surprising. `dots` should override the string. This also makes transitioning simpler.

Let's go with Michael's suggestion and syntax for duration objects, and deprecating everything else.

**Comment by 0xfe** [comment]
Also, I like the insight about linked vs. unlinked. My plan for dissimilar durations on a chord was to encode "override" durations on the contained `NoteHead` instances.

I'll need to think about this a bit more.

**Comment by mscuthbert** [comment]
In m21 we have separate chord and note objects, which is closer to a musician's interpretation but way harder in practice so I think that VexFlow's idea that a note is a special chord that happens to have one pitch is a good one.  However, along the way we realized that almost everything that a chord and a note had in common could be overridden for each pitch/notehead of a chord and (in a painful refactor) we changed chords so that they embedded an entire note object for each pitch. Then we set it up so that most of the attributes were by default shared among all pitches (via object references), so that using a setter/getter pair setting say a tie status for a chord set the ties on all the notes in the chord, but setting a tie on an individual note in the chord broke the relationship between that note and the chord's overall tie status.

Even if all Notes are Chords (StaveNotes) in VexFlow, something like this might become necessary at some point, letting the keys be complex objects in their own right that store information for that pitch. The instantiation wouldn't need to change but the object representation may.  Otherwise I found that I needed to keep lots of arrays of equal length where each element of each array represented some facet of the appearance of one pitch and every .add, .remove, etc. call had to be aware of each array, and it became hard to remember that when deleting a pitch its associated .noteheadParenthesisStatus array entry had to be removed as well.  Re-sorting pitches became the world's great pain-in-the-ass.

**Comment by aaronmars** [comment]
Sounds great to me!  I get this going.  Thanks for the insight, @mscuthbert

**Comment by 0xfe** [comment]
Michael, Agreed -- I definitely don't want to have parallel arrays for overrides. VexFlow never had separate `NoteHead` instances -- that was a relatively recent refactor.

We factored out `Stem`, `Flag`, and `NoteHead` from `StaveNote`, which now makes customizations and overrides (such as independent ticks) much simpler. And with the new SVG support for interactivity, you can attach events to the whole note, particular noteheads, just the stem, just the modifiers, or any arbitrary grouping.

**Comment by aaronmars** [comment]
While I'm refactoring the note specification stuff, I think it would be good to unpack the whole note duration string parsing thing.  In addition to dots being specified 2 ways (and adding a 3rd), there is also the issue of the `type` parameter.  This can be specified in 2 places as well.  Would it be good to deprecate the inline, parsed version?

Here's my suggestion for the One True Way of constructing a StaveNote:
```javascript
var n = new StaveNote({ keys: [ ... ], duration: { value: "4", dots: 2 }, type: "r" });
```
...where `keys` is unchanged from the current implementation, `duration.value` is required, `duration.dots` is optional (defaulting to 0), and `type` is how to specify one of 'n', 'r', 'h', 'm' or 's' (optional, defaulting to 'n').

Of course, there is a lot of backward compatibility to consider... especially when these things are defined simultaneously.  One would hope the existing test cases will cover that.

**Comment by 0xfe** [comment]
Obsolete.
